src/protocol/boolean/random_bits.rs

```diff
@@ -0,0 +1,170 @@
+use crate::error::BoxError;
+use crate::ff::{Field, Int};
+use crate::protocol::boolean::BitOpStep;
+use crate::protocol::modulus_conversion::convert_shares::{ConvertShares, XorShares};
+use crate::protocol::{context::ProtocolContext, RecordId};
+use crate::secret_sharing::Replicated;
+use futures::future::try_join_all;
+
+/// This protocol generates a sequence of uniformly random sharing of bits in `F_p`.
+/// Adding these 3-way secret-sharing will yield the secret `r_i ∈ {0,1}`.
+pub struct RandomBits {}
+
+impl RandomBits {
+    #[allow(dead_code)]
+    pub async fn execute<F: Field>(
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<(Vec<Replicated<F>>, Replicated<F>), BoxError> {
+        // We assume the bit length we operate in would not exceed 255
+        #[allow(clippy::cast_possible_truncation)]
+        let l = F::Integer::BITS as u8;
+
+        // Generate a pair of random numbers. We'll use these numbers as
+        // the source of `l`-bit long uniformly random sequence of bits.
+        let r_bits: (F, F) = ctx
+            .narrow(&Step::RandomFields)
+            .prss()
+            .generate_fields(record_id);
+
+        // Same here. For now, 64 bits is enough for our F_p
+        #[allow(clippy::cast_possible_truncation)]
+        let xor_shares = XorShares {
+            num_bits: l,
+            packed_bits_left: r_bits.0.as_u128() as u64,
+            packed_bits_right: r_bits.1.as_u128() as u64,
+        };
+
+        // Convert each bit to secret sharings of that bit in the target field
+        let c = ctx.narrow(&Step::ConvertShares);
+        let futures = (0..l).map(|i| {
+            // again, we don't expect our prime field to be > 2^64
+            #[allow(clippy::cast_possible_truncation)]
+            let c = c.narrow(&BitOpStep::Step(i as usize));
+            async move {
+                ConvertShares::new(xor_shares)
```

Comment:
Why are you instantiating this object `l` times?  It's the same object every time, isn't it?

Reply:
I pulled this out from the loop. It was because XorShare didn't implement Copy at first, but I added `derive`.